### PR TITLE
[MU4] Implemented native processing of frameless window in Windows

### PR DESCRIPTION
**Pull request by Eism**
- Aero snap is working now
- Moved the window resizing/drawing shadow to c++(to native event filter)

TODO: need implement for Linux too(or use the previous version on the frameless window)

https://user-images.githubusercontent.com/10116828/118992430-edb7cf00-b984-11eb-9af1-78c2f99f8df5.mp4

### CODE REVIEW

File: src/appshell/CMakeLists.txt [hunk 1]
```diff
@@ -87,9 +87,20 @@ set(MODULE_SRC
...
+    ${CMAKE_CURRENT_LIST_DIR}/view/framelesswindow/framelesswindowmodel.h
+    ${CMAKE_CURRENT_LIST_DIR}/view/framelesswindow/abstractnativeeventfilter.cpp
+    ${CMAKE_CURRENT_LIST_DIR}/view/framelesswindow/abstractnativeeventfilter.h
     ${DOCKWINDOW_SRC}
     )
 
+if (OS_IS_WIN)
+    set(MODULE_SRC ${MODULE_SRC}
+        ${CMAKE_CURRENT_LIST_DIR}/view/framelesswindow/windowsnativeeventfilter.cpp
+        ${CMAKE_CURRENT_LIST_DIR}/view/framelesswindow/windowsnativeeventfilter.h
```
Comment (igorkorsukov) [inline src/appshell/CMakeLists.txt:100]: I think it's better to move these files to `src/appshell/internal/platform/win`

File: src/appshell/qml/kdab/AppTitleBar.qml [hunk 2]
```diff
@@ -30,39 +30,34 @@ Rectangle {
 
     color: ui.theme.backgroundPrimaryColor
 
+    property var titleMoveAreaRect: {
+        return Qt.rect(titleMoveArea.x, titleMoveArea.y, titleMoveArea.width, titleMoveArea.height)
+    }
```
Comment (igorkorsukov) [inline src/appshell/qml/kdab/AppTitleBar.qml:35]: I think it better to do 
```
property var titleMoveAreaRect: Qt.rect(titleMoveArea.x, titleMoveArea.y, titleMoveArea.width, titleMoveArea.height)
```

File: src/appshell/view/framelesswindow/framelesswindowmodel.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+#ifdef Q_OS_WIN
+#include <dwmapi.h>
+#include "windowsnativeeventfilter.h"
+#else
+#include "abstractnativeeventfilter.h"
+#endif
+
+using namespace mu::appshell;
+
+FremelessWindowModel::FremelessWindowModel(QObject* parent)
```
Comment (igorkorsukov) [inline src/appshell/view/framelesswindow/framelesswindowmodel.cpp:36]: Frameless

File: src/appshell/view/framelesswindow/framelesswindowmodel.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    SetWindowLongPtr(hwnd, GWL_STYLE,
+                     static_cast<LONG>(WS_POPUP | WS_CAPTION | WS_SYSMENU | WS_MINIMIZEBOX | WS_MAXIMIZEBOX | WS_THICKFRAME
+                                       | WS_CLIPCHILDREN));
+
+    const MARGINS shadow_on = { 1, 1, 1, 1 };
+    DwmExtendFrameIntoClientArea(hwnd, &shadow_on);
+
+    SetWindowPos(hwnd, Q_NULLPTR, 0, 0, 0, 0, SWP_FRAMECHANGED | SWP_NOMOVE | SWP_NOSIZE);
+    ShowWindow(hwnd, SW_SHOW);
+#endif
```
Comment (igorkorsukov) [inline src/appshell/view/framelesswindow/framelesswindowmodel.cpp:67]: I think, better AbstractNativeEventFilter rename to FramelessWindowController and put this code to it.
qApp->installNativeEventFilter(m_nativeEventFilter); put to it too, so will be 
```
void FremelessWindowModel::init(QWindow* window)
{
    m_controller->init(window);
}
```

void WinFremelessWindowController::init(QWindow* window)
{
      qApp->installNativeEventFilter(this); 
      HWND hwnd = (HWND)window->winId();
      ...
}

Comment (Eism) [reply src/appshell/view/framelesswindow/framelesswindowmodel.cpp]: thanks! fixed

File: src/appshell/CMakeLists.txt (hunk 1 above)
Comment (Eism) [reply src/appshell/CMakeLists.txt]: thanks! fixed

**Review by RomanPudashkin — APPROVED** [review]

**Comment by vshalkevich** [comment]
Positioning works fine on Windows 10

**Pull request merged**